toolz/itertoolz.py

```diff
@@ -901,3 +903,45 @@ def peek(seq):
     iterator = iter(seq)
     item = next(iterator)
     return item, itertools.chain([item], iterator)
+
+
+def random_sample(prob, seq, random_state=None):
+    """
+    Return elements from a sequence with probability of prob
+
+    Returns a lazy iterator of random items from seq.
+
+    ``random_sample`` considers each item independently and without
+    replacement. See below how the first time it returned 13 items and the
+    next time it returned 6 items.
+
+    >>> seq = list(range(100))
+    >>> list(random_sample(0.1, seq)) # doctest: +SKIP
+    [6, 9, 19, 35, 45, 50, 58, 62, 68, 72, 78, 86, 95]
+    >>> list(random_sample(0.1, seq)) # doctest: +SKIP
+    [6, 44, 54, 61, 69, 94]
+
+    Providing an integer seed for ``random_state`` will result in
+    deterministic sampling. Given the same seed it will return the same sample
+    every time.
+
+    >>> list(random_sample(0.1, seq, random_state=2016))
+    [7, 9, 19, 25, 30, 32, 34, 48, 59, 60, 81, 98]
+    >>> list(random_sample(0.1, seq, random_state=2016))
+    [7, 9, 19, 25, 30, 32, 34, 48, 59, 60, 81, 98]
+
+    ``random_state`` can also be any object with a method ``random`` that
+    returns floats between 0.0 and 1.0 (exclusive).
+
+    >>> from random import Random
+    >>> randobj = Random(2016)
+    >>> list(random_sample(0.1, seq, random_state=randobj))
+    [7, 9, 19, 25, 30, 32, 34, 48, 59, 60, 81, 98]
+
+    """
+    if random_state is None or isinstance(random_state, int):
```

Comment:
The current version I'm playing with still has a type checker that looks like this:

```python
if random_state is None or isinstance(random_state,
                                      (Integral,
                                       collections.MutableSequence,
                                       collections.Sequence)):
    random_state = Random(random_state)
```

This will accept any integer types and will accept both byte and unicode strings. It also (appears) to have identical behaviour across 3.5, 2.7 and pypy-2, based on my somewhat limited tests.

It will also accept objects of other types, but this version still works well with passing an object with a ``random`` method. 

Next I'll play around with the idea of just passing anything in and working with the resulting errors.